src/components/views/elements/Dropdown.js

```diff
@@ -0,0 +1,323 @@
+        this._onMenuOptionClick = this._onMenuOptionClick.bind(this);
+        this._onInputKeyPress = this._onInputKeyPress.bind(this);
+        this._onInputKeyUp = this._onInputKeyUp.bind(this);
+        this._onInputChange = this._onInputChange.bind(this);
+        this._collectRoot = this._collectRoot.bind(this);
+        this._collectInputTextBox = this._collectInputTextBox.bind(this);
+        this._setHighlightedOption = this._setHighlightedOption.bind(this);
+
+        this.inputTextBox = null;
+
+        this._reindexChildren(this.props.children);
+
+        this.state = {
+            // True if the menu is dropped-down
+            expanded: false,
+            // The key of the selected option
+            selectedOption: this.props.initialSelectedOption || props.children[0].key,
+            // The key of the highlighted option
+            // (the option that would become selected if you pressed enter)
+            highlightedOption: props.children[0].key,
+            // the current search query
+            searchQuery: '',
+        };
+    }
+
+    componentWillMount() {
+        // Listen for all clicks on the document so we can close the
+        // menu when the user clicks somewhere else
+        document.addEventListener('click', this._onDocumentClick, false);
+
+        // fire this now so the defaults can be set up
+        this.props.onOptionChange(this.state.selectedOption);
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('click', this._onDocumentClick, false);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        this._reindexChildren(nextProps.children);
+        this.setState({
+            highlightedOption: Object.keys(this.childrenByKey)[0],
+        });
+    }
+
+    _reindexChildren(children) {
+        this.childrenByKey = {};
+        for (const child of children) {
+            this.childrenByKey[child.key] = child;
+        }
+    }
+
+    _onDocumentClick(ev) {
+        // Close the dropdown if the user clicks anywhere that isn't
+        // within our root element
+        if (ev !== this.ignoreEvent) {
+            this.setState({
+                expanded: false,
+            });
+        }
+    }
+
+    _onRootClick(ev) {
+        // This captures any clicks that happen within our elements,
+        // such that we can then ignore them when they're seen by the
+        // click listener on the document handler, ie. not close the
+        // dropdown immediately after opening it.
+        // NB. We can't just stopPropagation() because then the event
+        // doesn't reach the React onClick().
+        this.ignoreEvent = ev;
+    }
+
+    _onInputClick(ev) {
+        this.setState({
+            expanded: !this.state.expanded,
+        });
+        ev.preventDefault();
+    }
+
+    _onMenuOptionClick(dropdownKey) {
+        this.setState({
+            expanded: false,
+            selectedOption: dropdownKey,
```

Comment:
I think a more reacty way to do this is to make `selectedOption` a prop (instead of `initialSelectedOption`), and make the parent responsible for maintaining its value (in this case, you're already tracking it in the parent components' state so it would just be a case of `s/initial//` on the props)